src/meta/meta_bulk_load_service.cpp

```diff
@@ -283,41 +283,92 @@ void bulk_load_service::create_partition_bulk_load_dir(const std::string &app_na
-                         std::bind(&bulk_load_service::partition_bulk_load, this, app_name, pid),
+                         [this, retry_function, app_name, pid]() { retry_function(app_name, pid); },
+                         0,
+                         std::chrono::seconds(1));
+        return false;
+    }
+
+    if (pconfig.secondaries.size() < pconfig.max_replica_count - 1) {
+        bulk_load_status::type p_status;
+        {
+            zauto_read_lock l(_lock);
+            p_status = get_partition_bulk_load_status_unlocked(pid);
+        }
+        // rollback to downloading, pause,cancel,failed bulk load should always send to replica
+        // server
+        if (!always_unhealthy_check && (p_status == bulk_load_status::BLS_DOWNLOADING ||
+                                        p_status == bulk_load_status::BLS_PAUSING ||
+                                        p_status == bulk_load_status::BLS_CANCELED ||
+                                        p_status == bulk_load_status::BLS_FAILED)) {
+            return true;
+        }
+        dwarn_f("app({}) partition({}) is unhealthy, status({}), try it later",
+                app_name,
+                pid,
+                dsn::enum_to_string(p_status));
+        tasking::enqueue(LPC_META_STATE_NORMAL,
+                         _meta_svc->tracker(),
+                         [this, retry_function, app_name, pid]() { retry_function(app_name, pid); },
                          0,
                          std::chrono::seconds(1));
+        return false;
+    }
+    return true;
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::partition_bulk_load(const std::string &app_name, const gpid &pid)
+{
+    FAIL_POINT_INJECT_F("meta_bulk_load_partition_bulk_load", [](dsn::string_view) {});
+
+    partition_configuration pconfig;
+    if (!check_partition_status(app_name,
+                                pid,
+                                false,
+                                std::bind(&bulk_load_service::partition_bulk_load,
+                                          this,
+                                          std::placeholders::_1,
+                                          std::placeholders::_2),
+                                pconfig)) {
         return;
     }
 
-    zauto_read_lock l(_lock);
-    const app_bulk_load_info &ainfo = _app_bulk_load_info[pid.get_app_id()];
+    rpc_address primary_addr = pconfig.primary;
     auto req = make_unique<bulk_load_request>();
-    req->pid = pid;
-    req->app_name = app_name;
-    req->primary_addr = primary_addr;
-    req->remote_provider_name = ainfo.file_provider_type;
-    req->cluster_name = ainfo.cluster_name;
-    req->meta_bulk_load_status = get_partition_bulk_load_status_unlocked(pid);
-    req->ballot = app->partitions[pid.get_partition_index()].ballot;
-    req->query_bulk_load_metadata = is_partition_metadata_not_updated_unlocked(pid);
-    req->remote_root_path = ainfo.remote_root_path;
+    {
```

Comment:
I think there is no need to add lock for `req`

Reply:
The `_lock` is not for the req, but is used for following:
```
const app_bulk_load_info &ainfo = _app_bulk_load_info[pid.get_app_id()];
req->meta_bulk_load_status = get_partition_bulk_load_status_unlocked(pid);
req->query_bulk_load_metadata = is_partition_metadata_not_updated_unlocked(pid);
```
The lock is added for `_app_bulk_load_info` and two functions, all of them is used to set parameter for `req`. To make code clear, I add the lock as pull request shows.

Reply:
Maybe it's better to wrap a thread-safe map here, it's easier to read the codes. Just keep these codes in this PR